packages/salesforcedx-vscode-replay-debugger/src/states/variableAssignmentState.ts

```diff
@@ -26,106 +26,128 @@ export class VariableAssignmentState implements DebugLogState {
+                pulledRef.name = varName;
+                pulledRef.variablesRef = logContext
+                  .getVariableHandler()
+                  .create(pulledRef);
+                refContainer.variables.set(varName, pulledRef);
+              } else {
+                refContainer.variables.set(
+                  varName,
+                  new ApexVariableContainer(varName, value, '')
+                );
+              }
+            }
+          } else if (value.indexOf('{') === 0) {
+            this.parseJSONAndPopulate(value, refContainer, logContext);
+          } else {
+            refContainer.variables.set(
+              varName,
+              new ApexVariableContainer(varName, value, '')
+            );
+          }
         }
-        // assigning to variable's fields, currently only going to work for a variable in local
-      } else if (name.indexOf('.') !== -1 && ref !== '0') {
-        container = logContext.getRefsMap().get(ref)!;
-        if (container) {
+
+        // update the correct toplevel container
+        if (
+          container &&
+          !container.type.startsWith('Map<') &&
+          !container.type.startsWith('List<') &&
+          !container.type.startsWith('Set<')
+        ) {
+          container.ref = ref;
           container.value = '';
+          container.variables = refContainer.variables;
+          refContainer.type = container.type;
           if (container.variablesRef === 0) {
             container.variablesRef = logContext
               .getVariableHandler()
               .create(container);
+          } else if (value === '{}') {
+            container.value = value;
           }
-          if (value.indexOf('{') !== -1 && value !== '{}') {
-            const topLevel = new ApexVariableContainer(varName, '', '');
-            container.variables.set(varName, topLevel);
-            topLevel.variablesRef = logContext
-              .getVariableHandler()
-              .create(topLevel);
-            this.parseAndPopulate(value, topLevel);
-          } else {
-            container.variables.set(
-              varName,
-              new ApexVariableContainer(varName, value, '')
-            );
-          }
+        }
+      } else {
+        if (container) {
+          container.value = value;
         }
       }
     }
     return false;
   }
 
-  private parseAndPopulate(value: string, container: ApexVariableContainer) {
+  private parseJSONAndPopulate(
+    value: string,
+    container: ApexVariableContainer,
+    logContext: LogContext
+  ) {
     try {
       const obj = JSON.parse(value);
       Object.keys(obj).forEach(key => {
-        container.variables.set(
-          key,
-          new ApexVariableContainer(key, String(obj[key]), '')
-        );
+        if (logContext.getRefsMap().has(String(obj[key]))) {
+          const refContainer = logContext.getRefsMap().get(String(obj[key]))!;
```

Comment:
Here is the other location where we check for a previously seen memory reference. This is in case we get a json value where key value pairs lead to an address

Reply:
Does using `as string` work here? Also, it looks like you can cast to string once right after the forEach then re-use the string constant below.

Reply:
James, I'd previously asked you about this. Map.get is going to return undefined, instead of making a call to .has and then a call to .get, effectively double tapping the map, we should be able to make a call to get and check if refContainer is defined.